File: Python/Tests/cliquet_option.py
"""
/*
 * Copyright (C) 2001 QuantLib Group
 * 
 * This file is part of QuantLib.
 * QuantLib is a C++ open source library for financial quantitative
 * analysts and developers --- http://quantlib.sourceforge.net/
 *
 * QuantLib is free software and you are allowed to use, copy, modify, merge,
 * publish, distribute, and/or sell copies of it under the conditions stated 
 * in the QuantLib License.
 *
 * This program is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You should have received a copy of the license along with this file;
 * if not, contact ferdinando@ametrano.net
 *
 * QuantLib license is also available at:
 * http://quantlib.sourceforge.net/LICENSE.TXT
*/
"""
""" 
    $Source$
    $Log$
    Revision 1.2  2001/04/04 11:08:11  lballabio
    Python tests implemented on top of PyUnit

    Revision 1.1  2001/04/04 10:01:36  marmar
    introducing cliquet option

"""

import QuantLib
import unittest
from math import exp

class CliquetOptionTest(unittest.TestCase):
    def runTest(self):
        "Testing cliquet option pricer"
        spot = 100
        divYield = 0.01
        rRate = 0.06
        dates = [0.50, 1.00, 2.00]
        vol = 0.35
        euro1 = QuantLib.BSMEuropeanOption("Call", spot, spot, divYield,
                    rRate, dates[1]-dates[0], vol)
        w1 = exp(divYield*dates[0])
        
        euro2 = QuantLib.BSMEuropeanOption("Call", spot, spot, divYield,
                    rRate, dates[2]-dates[1], vol)
        w2 = exp(divYield*dates[1])
         
        cliquet = QuantLib.CliquetOption("Call", spot, divYield, rRate,
                      dates, vol)
        for method in ['value', 'delta', 'gamma', 'theta', 'rho', 'vega']:
            euro1Method = getattr(euro1, method)
            euro2Method = getattr(euro2, method)
            cliquetMethod = getattr(cliquet, method)
            expected = w1*apply(euro1Method,()) + w2 * apply(euro2Method,())
            calculated = apply(cliquetMethod,())
            error = abs(calculated-expected)
            assert error <= 1e-10, \
                'wrong %s\n' % method + \
                'calculated: %f\n' % calculated + \
                'expected  : %f\n' % expected


if __name__ == '__main__':
    suite = unittest.TestSuite()
    suite.addTest(CliquetOptionTest())
    unittest.TextTestRunner().run(suite)

File: Sources/Pricers/cliquetoption.cpp

/*
 * Copyright (C) 2000-2001 QuantLib Group
 *
 * This file is part of QuantLib.
 * QuantLib is a C++ open source library for financial quantitative
 * analysts and developers --- http://quantlib.org/
 *
 * QuantLib is free software and you are allowed to use, copy, modify, merge,
 * publish, distribute, and/or sell copies of it under the conditions stated
 * in the QuantLib License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You should have received a copy of the license along with this file;
 * if not, please email quantlib-users@lists.sourceforge.net
 * The license is also available at http://quantlib.org/LICENSE.TXT
 *
 * The members of the QuantLib Group are listed in the Authors.txt file, also
 * available at http://quantlib.org/group.html
*/

/*!  \file cliquetoption.cpp

    \fullpath
    Sources/Pricers/%cliquetoption.cpp
   \brief Textbook example of european-style multi-period option.

*/

// $Id$
// $Log$
// Revision 1.13  2001/08/31 15:23:47  sigmud
// refining fullpath entries for doxygen documentation
//
// Revision 1.12  2001/08/09 14:59:48  sigmud
// header modification
//
// Revision 1.11  2001/08/08 11:07:50  sigmud
// inserting \fullpath for doxygen
//
// Revision 1.10  2001/08/07 11:25:55  sigmud
// copyright header maintenance
//
// Revision 1.9  2001/08/06 15:43:34  nando
// BSMOption now is SingleAssetOption
// BSMEuropeanOption now is EuropeanOption
//
// Revision 1.8  2001/07/25 15:47:29  sigmud
// Change from quantlib.sourceforge.net to quantlib.org
//
// Revision 1.7  2001/05/24 15:40:10  nando
// smoothing #include xx.hpp and cutting old Log messages
//

#include "ql/Pricers/cliquetoption.hpp"
#include "ql/Math/normaldistribution.hpp"

namespace QuantLib
{
    namespace Pricers
    {
        CliquetOption::CliquetOption(Type type,
                                     double underlying,
                                     Rate dividendYield,
                                     Rate riskFreeRate,
                                     const std::vector<Time> &dates,
                                     double volatility)
        : SingleAssetOption(type, underlying, underlying, dividendYield,
                    riskFreeRate, dates[dates.size()-1], volatility),
        numPeriods_(dates.size()-1),
        optionlet_(numPeriods_),
        weight_(numPeriods_){

            QL_REQUIRE(numPeriods_ >= 1,
                       "At least two dates are required for cliquet options");

            for(int i = 0; i < numPeriods_; i++){
                weight_[i] = QL_EXP(dividendYield * dates[i]);
                optionlet_[i] = Handle<EuropeanOption>(
                    new EuropeanOption(type,
                                          underlying,
                                          underlying,
                                          dividendYield,
                                          riskFreeRate,
                                          dates[i+1] - dates[i],
                                          volatility));
            }

        }

        Handle<SingleAssetOption> CliquetOption::clone() const {
            return Handle<SingleAssetOption>(new CliquetOption(*this));
        }

        double CliquetOption::value() const {
            double optValue = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optValue += weight_[i] * optionlet_[i] -> value();
            return optValue;
        }

        double CliquetOption::delta() const {
            double optDelta = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optDelta += weight_[i] * optionlet_[i] -> delta();
            return optDelta;
        }

        double CliquetOption::gamma() const {
            double optGamma = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optGamma += weight_[i] * optionlet_[i] -> gamma();
            return optGamma;
        }

        double CliquetOption::theta() const {
            double optTheta = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optTheta += weight_[i] * optionlet_[i] -> theta();
            return optTheta;
        }

        double CliquetOption::rho() const {
            double optRho = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optRho += weight_[i] * optionlet_[i] -> rho();
            return optRho;
        }

        double CliquetOption::vega() const {
            double optVega = 0.0;
            for(int i = 0; i < numPeriods_; i++)
                optVega += weight_[i] * optionlet_[i] -> vega();
            return optVega;
        }

    }

}







